hack/test-go.sh

```diff
@@ -1,129 +1,259 @@
-    # if the package name ends with /... they are intending a recursive test
-    test_packages=$(find_test_dirs "${1%$suffix}")
-  else
-    test_packages="$OS_GO_PACKAGE/$1"
-  fi
-elif [ -n "$TEST_KUBE" ]; then
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs; find_upstream_test_dirs`
+# check to see if user has not disabled coverage mode
+if [[ -n "${coverage_spec}" ]]; then
+    # if we have a coverage spec set, we add it. '-race' implies '-cover -covermode atomic'
+    # but specifying both at the same time does not lead to an error so we can add both specs
+    gotest_flags="${gotest_flags} ${coverage_spec}"
+fi
+
+# check to see if user has not disabled test timeouts
+if [[ -n "${test_timeout}" ]]; then
+    gotest_flags="${gotest_flags} -timeout ${test_timeout}"
+fi
+
+# list_test_packages_under lists all packages containing Golang test files that we want to run as unit tests
+# under the given base dir in the OpenShift Origin tree
+function list_test_packages_under() {
+    local basedir=$@
+
+    # we do not quote ${basedir} to allow for multiple arguments to be passed in as well as to allow for
+    # arguments that use expansion, e.g. paths containing brace expansion or wildcards
+    find ${basedir} -not \(                   \
+        \(                                    \
+              -path 'Godeps'                  \
+              -o -path '*_output'             \
+              -o -path '*_tools'              \
+              -o -path '*.git'                \
+              -o -path '*openshift.local.*'   \
+              -o -path '*Godeps/*'            \
+              -o -path '*assets/node_modules' \
+              -o -path '*test/*'              \
+        \) -prune                             \
+    \) -name '*_test.go' -print0 | xargs -0n1 dirname -z | sort -uz | xargs -0n1 printf "${OS_GO_PACKAGE}/%s\n"
+}
+
+# Break up the positional arguments into packages that need to be tested and arguments that need to be passed to `go test`
+first_flag_arg=-1
+for (( i = 1; i < "$#"; i++ )); do
+    if [[ "${!i:0:1}" = "-" ]]; then
+        # we found an arg that begins with a dash, so we stop interpreting arguments 
+        # henceforth as packages and instead interpret them as flags to give to `go test`
+        first_flag_arg=$i
+        break
+    fi
+done
+
+package_args=
+if [[ ${first_flag_arg} -gt 0 ]]; then
+    # we recieved some `go test` flags as args, so we add them to the gotest_flags
+    gotest_flags="${gotest_flags} ${@:${first_flag_arg}}"
+
+    # the other args aside from $0 are packages to test
+    package_args="${@:1:${first_flag_arg}-1}"
 else
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs`
+    # we got no flags as args, so all args are packages to test
+    package_args="$@"
 fi
 
-if [ -n "$PRINT_PACKAGES" ]; then
-  for package in $test_packages
-  do
-    echo $package
-  done
+# Determine packages to test
+test_packages=
+if [[ "${#package_args}" -gt 0 ]]; then
```

Comment:
This should work, but because `$package_args` is a string (not an array), it seems like `[[ -n "$package_args" ]]` would be more natural.